src/functions/routers/mod.ts

```diff
@@ -0,0 +1,125 @@
+import {
+  ParameterSetDefinition,
+  PossibleParameterKeys,
+} from "../../parameters/mod.ts";
+import { SlackFunction } from "../mod.ts";
+import type { ActionHandler } from "./types.ts";
+
+/**
+ * Define an actions "router" and its input and output parameters for use in a Slack application. The ActionsRouter will route incoming action events to action-specific handlers.
+ * @param {SlackFunction<InputParameters, OutputParameters, RequiredInput, RequiredOutput>} func Reference to your previously-created SlackFunction, defined via DefineFunction
+ * @returns {ActionsRouter}
+ */
+export const BlockActionsRouter = <
+  InputParameters extends ParameterSetDefinition,
+  OutputParameters extends ParameterSetDefinition,
+  RequiredInput extends PossibleParameterKeys<InputParameters>,
+  RequiredOutput extends PossibleParameterKeys<OutputParameters>,
+>(
+  func: SlackFunction<
+    InputParameters,
+    OutputParameters,
+    RequiredInput,
+    RequiredOutput
+  >,
+) => {
+  return new ActionsRouter(func);
+};
+
+class ActionsRouter<
+  InputParameters extends ParameterSetDefinition,
+  OutputParameters extends ParameterSetDefinition,
+  RequiredInput extends PossibleParameterKeys<InputParameters>,
+  RequiredOutput extends PossibleParameterKeys<OutputParameters>,
+> {
+  private func: SlackFunction<
+    InputParameters,
+    OutputParameters,
+    RequiredInput,
+    RequiredOutput
+  >;
+  private routes: Array<
+    Array<string[] | RegExp | ActionHandler<typeof this.func.definition>>
+  >;
+
+  constructor(
+    func: SlackFunction<
+      InputParameters,
+      OutputParameters,
+      RequiredInput,
+      RequiredOutput
+    >,
+  ) {
+    this.func = func;
+    this.routes = [];
+  }
+
+  /**
+   * Add an action handler for something that can matche an action_id.
+   */
+  add(
+    action_id: string | string[] | RegExp,
+    handler: ActionHandler<typeof this.func.definition>,
+  ): ActionsRouter<
+    InputParameters,
+    OutputParameters,
+    RequiredInput,
+    RequiredOutput
+  > {
+    if (typeof action_id === "string") {
+      action_id = [action_id];
+    }
+    this.routes.push([action_id, handler]);
+    return this;
+  }
+
+  /**
+   * Returns a method handling routing of action payloads to the appropriate action handler.
+   * The output of export() should be attached to the `blockActions` export of your function.
+   */
+  export(): ActionHandler<typeof this.func.definition> {
+    return async (context) => {
+      const action = context.action.action_id as string;
+      const handler = this.matchHandler(action);
```

Comment:
I think we should expand what our criteria matching handles to both `action_id` and `block_id`. This is the route bolt ended up going, and it ended up satisfying a lot of use-cases. For example, being able to do the following:


### 2 basic ways to call

First way: this assumes arg is an action_id
```ts
router.add(action_id: string, ...)
```

Second Way: This allows for more complex matching against block_ids and action_ids
```ts
router.add(criteria: { block_id?: string, action_id?: string }, ...)
```

Then we can expand where we accept a `string` to accept `regex`, or an `array` of string/regex

### Examples:

When arg is just a string, just matches on action_id (the most common scenario)
```ts
router.add('my_action_id'), ...);
```

Matches on just block_id 
```ts
router.add({ block_id: 'my_block_id'}, ...)
```

Matches on both block_id and action_id
```ts
router.add({ block_id: 'my_block_id', action_id: 'my_action_id'}, ...)
```

Matches on just action_id - mostly for consistency w/ the interface, but you'd probably use the plain string arg for just action_id
```ts
router.add({ action_id: 'my_block_id'}, ...)
```

#### Arrays
This feels like we could defer for v1 and add as a followup unless it's easier to just do now

Match on multiple action ids
```ts
router.add(['action1', 'action2'), ...);
```

Also support arrays in all the other scenarios
```ts
router.add({ block_id: ['block1', 'block2']}, ...)
```
This one feels more complicated, and perhaps less useful but for consistency in the interface would be nice to support, would just match if any block_ids match, and any action_ids match
```ts
router.add({ block_id: ['block1', 'block2'], action_id: ['action1', 'action2']}, ...)
```

###Regex
We could also defer regex and start w/ just strings, but eventually use a regex anywhere a string can be used too
```ts
router.add(/^my_prefix/, ...)
router.add({ block_id: /^my_blocks/, action_id: /_id$/}, ...)
// also support arrays w/ regexes
```

Reply:
My suggestion would be we follow the path paved from bolt js and make it an AND. From an example on the bolt js docs:

```js
// Your listener function will only be called when the action_id matches 'select_user' AND the block_id matches 'assign_ticket'
app.action({ action_id: 'select_user', block_id: 'assign_ticket' },

```

We're adding support for arrays here, which bolt doesn't have, but I think the AND between block_id and action_id can behave the same way as bolt. What do you think?

Reply:
Ya that sounds good, just will be tricky to communicate that but that can be shelved as a documentation problem 😄 
The way I'm understand it is: logical OR for matching to a specific, single property, but logical AND _across_  all properties of complex constraint objects.